src/Http/ActionDispatcher.php

```diff
@@ -79,22 +77,15 @@ protected function _invoke(Controller $controller): ResponseInterface
         }
 
         $response = $controller->invokeAction();
-        if ($response !== null && !($response instanceof ResponseInterface)) {
-            throw new LogicException('Controller actions can only return Cake\Http\Response or null.');
-        }
-
-        if (!$response && $controller->isAutoRenderEnabled()) {
+        if ($response === null && $controller->isAutoRenderEnabled()) {
             $controller->render();
         }
 
         $result = $controller->shutdownProcess();
         if ($result instanceof ResponseInterface) {
             return $result;
         }
-        if (!$response) {
```

Comment:
We dont need this part anymore? We now always fetch from $controller->getResponse()?

Reply:
We don't, since return response returned by action is also assigned to `Controller::$response`. Plus beforeRender() / afterFilter() could have modified `Controller::$response` so that what holds the reference to most up to date response instance.